### ISSUE TITLE
Think about storing settings in a different format (instead of JSON)

### ISSUE THREAD

**Issue by LukasKalbertodt** [elan-ev/opencast-studio#535]
We plan on adding quite a few configuration options still, in particular make a few texts configurable. Large configurations with long texts probably won't look too bad in JSON, making the whole file hard to read. JSON also does not allow comments. So it might be worth thinking about an alternative format. 

YAML was mentioned. I would probably prefer TOML as it's a very minimal format, whereas YAML has a huge grammar and stuff. A simpler and smaller format is better for humans and beneficial as the parsing library will probably be smaller.

Here is an example configuration in different formats:

**JSON**
```json
{
  "opencast": {
    "serverUrl": "/",
    "loginProvided": true
  },
  "upload": {
    "workflowId": "fast"
  },
  "display": {
    "maxFps": 30
  },
  "texts": {
    "uploadSuccessful": {
      "en": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.",
      "de": "Loremipsum dolorsit amet, consectetur adipiscingelit, seddo eiusmod tempor incididunt utlabore etdolore magnaaliqua. Utenim ad minim veniam, quis nostrud exercitationullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eufugiat nulla pariatur."
    },
    "firstVisit": {
      "en": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
      "de": "Loremipsum dolorsit amet, consectetur adipiscingelit, seddo eiusmod tempor incididunt utlabore etdolore magnaaliqua. Utenim ad minim veniam, quis nostrud exercitationullamco laboris nisi ut aliquip ex ea commodo consequat."
    }
  }
}
```

**TOML**
```toml
[opencast]
serverUrl = "/"
loginProvided = true

[upload]
# A comment :)
workflowId = "fast"

[display]
maxFps = 30

[texts.uploadSuccessful]
en = """
Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor 
incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis 
nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. 
Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu 
fugiat nulla pariatur.
"""
de = """
Loremipsum dolorsit amet, consectetur adipiscingelit, seddo eiusmod 
tempor incididunt utlabore etdolore magnaaliqua. Utenim ad minim veniam, quis 
nostrud exercitationullamco laboris nisi ut aliquip ex ea commodo consequat. 
Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore 
eufugiat nulla pariatur.
"""

[texts.firstVisit]
# Long single line texts are allowed too, of course
en = "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat."
de = "Loremipsum dolorsit amet, consectetur adipiscingelit, seddo eiusmod tempor incididunt utlabore etdolore magnaaliqua. Utenim ad minim veniam, quis nostrud exercitationullamco laboris nisi ut aliquip ex ea commodo consequat."
```

**YAML**

```yaml
---
# Connection information for Opencast
# Setting these here will disable the settings in the Studio user interface
opencast:

  # URL of where to find Opencast for uploads. 
  # If Studio lives on the same server, a path will do
  serverUrl: "/"

  # If you already logged in to Opencast when you open Studio (e.g. since Studio is included in Opencast).
  # Setting this to false will make Studio ask the user for credentials in its settings.
  # Default: false
  loginProvided: true

# …
upload:
  # The workflow to use for ingests to Opencast
  # Default: studio-upload
  #workflowId: studio-upload

display:
  maxFps: 30

texts:
  uploadSuccessful:
    en: Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor
      incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud
      exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute
      irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat
      nulla pariatur.
    de: Loremipsum dolorsit amet, consectetur adipiscingelit, seddo eiusmod tempor
      incididunt utlabore etdolore magnaaliqua. Utenim ad minim veniam, quis nostrud
      exercitationullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute
      irure dolor in reprehenderit in voluptate velit esse cillum dolore eufugiat
      nulla pariatur.
  firstVisit:
    en: Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor
      incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud
      exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
    de: Loremipsum dolorsit amet, consectetur adipiscingelit, seddo eiusmod tempor
      incididunt utlabore etdolore magnaaliqua. Utenim ad minim veniam, quis nostrud
      exercitationullamco laboris nisi ut aliquip ex ea commodo consequat.
```

**Comment by mtneug** [comment]
Personally, I don't care much which of the three are used (although comments would be nice), though currently, YAML and JSON are used for other front-end tools. Opencast backend configuration is typically using Java properties (and XML of course, but let's not talk about that), which TOML is a kind of a more powerful version of.

**Comment by lkiesow** [comment]
Opencast already uses XML, Java properties, JSON and YAML. There is no sane reason to go for yet another format. It's not making things any easier for humans since you are forcing them to learn yet another thing for solving the exact same problem.

In addition to that, I think YAML is far easier to read. But that might as well be personal since I work more with YAML.

**Comment by LukasKalbertodt** [comment]
> There is no sane reason to go for yet another format.

But what if it's [better](https://xkcd.com/927/)?!?!

Jokes aside, I don't have too strong of an opinion on this one. And I am biased towards TOML just because I use it more often. I just wanted to bring up another alternative and maybe spark a discussion.

But for the protocol, there are good reasons not to use YAML. See [this post](https://www.arp242.net/yaml-config.html) for example. Some points: YAML is super complex, only easily human-readable when only using a small subset and often has surprising behavior (in particular regarding non-quoted strings). Also, comparing the sizes of the most popular [TOML JS library](https://bundlephobia.com/result?p=toml@3.0.0) (27.5kb minified, 6kb minified+gzipped) to the most popular [YAML JS library](https://bundlephobia.com/result?p=js-yaml@3.13.1) (171.9kb minified, 41.3kb minified+gzipped)... that's a clear win for TOML. 

Just throwing that out there ;-)

**Comment by JulianKniephoff** [comment]
> [...] since you are forcing them to learn yet another thing for solving the exact same problem.

That **can** be a very sane reason to add something.

I like YAML very much as a format. However, I also already got bitten by some of it's more surprising behaviors at least twice now, and it wasn't even when I was using the more arcane features which I also actually like a lot but got quite some flack for in the office when I did. :wink:

Long story short, I do like YAML, but I can see a place for TOML. Properties-files don't have enough build in structure IMO and JSON is not a configuration format, period.
